internal/runners/uploadingredient/uploadingredient.go

```diff
@@ -0,0 +1,343 @@
+	Name         string
+	Version      string
+	Namespace    string
+	Description  string
+	Authors      captain.UsersFlag
+	Depends      captain.PackagesFlag
+	Filepath     string
+	MetaFilepath string
+	Edit         bool
+	Editor       bool
+}
+
+type Runner struct {
+	auth    *auth.Auth
+	out     output.Outputer
+	prompt  prompt.Prompter
+	project *project.Project
+	client  *gqlclient.Client
+}
+
+type primeable interface {
+	primer.Outputer
+	primer.Auther
+	primer.Projecter
+	primer.Prompter
+}
+
+func New(prime primeable) *Runner {
+	client := gqlclient.NewWithOpts(
+		api.GetServiceURL(api.ServiceBuildPlan).String(), 0,
+		graphql.WithHTTPClient(http.DefaultClient),
+		graphql.UseMultipartForm(),
+	)
+	client.EnableDebugLog()
+	client.SetTokenProvider(prime.Auth())
+	client.EnableDebugLog()
+	return &Runner{auth: prime.Auth(), out: prime.Output(), prompt: prime.Prompt(), project: prime.Project(), client: client}
+}
+
+func (r *Runner) Run(params *Params) error {
+	if !r.auth.Authenticated() {
+		return locale.NewInputError("err_auth_required")
+	}
+
+	if !fileutils.FileExists(params.Filepath) {
+		return locale.NewInputError("err_uploadingredient_file_not_found", "File not found: {{.V0}}", params.Filepath)
+	}
+
+	reqVars := request.PublishVariables{}
+
+	// Pass input from meta file
+	if params.MetaFilepath != "" {
+		if !fileutils.TargetExists(params.MetaFilepath) {
+			return locale.NewInputError("err_uploadingredient_file_not_found", "File not found: {{.V0}}", params.MetaFilepath)
+		}
+
+		b, err := fileutils.ReadFile(params.MetaFilepath)
+		if err != nil {
+			return locale.WrapInputError(err, "err_uploadingredient_file_read", "Could not read file: {{.V0}}", params.MetaFilepath)
+		}
+
+		if err := yaml.Unmarshal(b, &reqVars); err != nil {
+			return locale.WrapInputError(err, "err_uploadingredient_file_read", "Failed to unmarshal meta file, error received: {{.V0}}", err.Error())
+		}
+	}
+
+	// Namespace
+	if params.Namespace != "" {
+		reqVars.Namespace = params.Namespace
+	} else if reqVars.Namespace == "" {
+		reqVars.Namespace = model.NewSharedNamespace(r.project.Owner()).String()
+	}
+
+	// Validate namespace
+	prefix := r.project.Owner() + "/"
+	if !strings.HasPrefix(reqVars.Namespace, prefix) {
+		return locale.NewInputError("err_uploadingredient_namespace_invalid_org",
+			"Namespace should be prefixed '[ACTIONABLE]{{.V0}}[/RESET]', received: '[ACTIONABLE]{{.V1}}[/RESET]'.", prefix, reqVars.Namespace)
+	}
+
+	// Name
+	if params.Name != "" { // Validate & Set name
+		reqVars.Name = params.Name
+	} else if reqVars.Name == "" {
+		reqVars.Name = filepath.Base(params.Filepath)
+	}
+
+	ts := time.Now()
+	ingredients, err := model.SearchIngredientsStrict(reqVars.Namespace, reqVars.Name, true, false, &ts)
+	if err != nil && !errs.Matches(err, &model.ErrSearch404{}) { // 404 means namespace not found, which is fine
+		return locale.WrapError(err, "err_uploadingredient_search", "Could not search for ingredient")
+	}
```

Comment:
If this is done to check for a namespace, it feels super janky and like a story should be filed somewhere for a better solution. If it's to gather ingredients to edit one, taking the first in the slice seems possibly concerning. Also, can this chunk be moved inside of the `if params.Edit` block?

Reply:
I don't follow how this is janky. Not finding any ingredients is a legitimate result that can happen, and a 404 is what this API responds with in that case.

Moving it inside the edit conditional isn't an option, as the result is used in the "else" block of that conditional.

Reply:
https://softwareengineering.stackexchange.com/a/203493

Reply:
That's because when the API was developed the namespaces were predetermined, and so supplying a namespace that didn't exist was likely a fair case for a 404.

Now that namespaces have become more variable we could probably revisit that, but then we also need to consider backward compatibility.

Frankly I just don't see this being worth the time investment. It might not fully respect the standards that are out there, but few API's do and spending time JUST to satisfy a standard is in my opinion a waste.